### PR TITLE
[ticket/16690] Fix htmlspecialchars and htmlspecialchars_decode default flag

### PR DESCRIPTION
**Pull request by rxu**
Checklist:

- [x] Correct branch: master for new features; 3.3.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html) and [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

<a href="https://tracker.phpbb.com/browse/PHPBB3-16690">PHPBB3-16690</a>.

### CODE REVIEW

File: phpBB/includes/acp/acp_attachments.php [hunk 1]
```diff
@@ -759,7 +759,7 @@ function main($id, $mode)
 									continue;
 								}
 
-								$filename_list .= '<option value="' . htmlspecialchars($img) . '"' . $selected . '>' . htmlspecialchars($img) . '</option>';
+								$filename_list .= '<option value="' . htmlspecialchars(, ENT_COMPAT) . '"' . $selected . '>' . htmlspecialchars($img, ENT_COMPAT) . '</option>';
```
Comment (CHItA) [inline phpBB/includes/acp/acp_attachments.php:762]: `$img` went missing.

**Review by CHItA — APPROVED** [review]

**Comment by rxu** [comment]
There's a problem with convertor which has entries like
```php
array('poll_title',				'vote_desc.vote_text',				array('function1' => 'null_to_str', 'function2' => 'phpbb_set_encoding', 'function3' => 'htmlspecialchars_decode', 'function4' => 'utf8_htmlspecialchars')),
```
where `htmlspecialchars_decode` is used with no option to change the flag.

**Comment by rxu** [comment]
Also I'm not sure if absolutely all entries of those functions should have `ENT_COMPAT` flag by default. See https://news-web.php.net/php.internals/112782

**Comment by CHItA** [comment]
> ```
> htmlspecialchars_decode
> ```

There are two places where that is used: [here](https://github.com/phpbb/phpbb/blob/master/phpBB/install/convert/convertor.php#L1470) and [here](https://github.com/phpbb/phpbb/blob/master/phpBB/install/convert/convertor.php#L1520).

You can hard code what you want at those places. As for using `ENT_COMPAT` everywhere, it seems fine. Most of these cases should be managed in/by Twig. I wouldn't worry about it too much, as this was the behavior so far it should be fine. If you find any places where something else work better, feel free to update them. I just don't think that is required.

File: phpBB/install/convert/convertor.php [hunk 2]
```diff
@@ -1468,6 +1468,12 @@ function process_row(&$schema, &$sql_data, &$insert_values)
 									$value = array($value);
 								}
 
+								// Add ENT_COMPAT default flag to html specialchars/entities functions, see PHPBB3-16690
+								if (in_array($execution, ['htmlspecialchars', 'htmlentities', 'htmlspecialchars_decode', 'html_entitity_decode']))
+								{
+									$value[] = 'ENT_COMPAT';
```
Comment (CHItA) [inline phpBB/install/convert/convertor.php:1474]: I think this should be the constant. Not a string.

**Pull request merged**